src/exporters/mod.rs

```diff
@@ -5,54 +5,142 @@ mod stdout;
 use std::io;
 use std::{thread, time::Duration};
 
+extern crate rand;
 use crossbeam_channel::Receiver;
+use rand::Rng;
 
 use crate::message::Message;
 pub use influxdb::InfluxDB;
 pub use stdout::Stdout;
 
+const ZERO: Duration = Duration::from_secs(0);
+
 pub struct ExporterScheduler {
-    exporter: Option<Box<dyn Exporter + Send>>,
+    exporter: Box<dyn Exporter + Send>,
+    receiver: Receiver<Message>,
+    retry_buffer: RetryItem,
 }
 
 impl ExporterScheduler {
-    pub fn new() -> ExporterScheduler {
-        ExporterScheduler { exporter: None }
-    }
-
-    pub fn register(&mut self, exporter: impl Exporter + Send + 'static) {
-        self.exporter.replace(Box::new(exporter));
+    pub fn new(
+        exporter: impl Exporter + Send + 'static,
+        receiver: Receiver<Message>,
+    ) -> ExporterScheduler {
+        ExporterScheduler {
+            exporter: Box::new(exporter),
+            receiver,
+            retry_buffer: RetryItem::new(Duration::from_secs(120)),
+        }
     }
 
-    pub fn start(&self, receiver: Receiver<Message>) {
-        if self.exporter.is_none() {
-            println!("No exporters configured!");
-            return;
-        }
+    pub fn start(&mut self) {
         println!("Exporter scheduler started");
 
         loop {
-            let msg = match receiver.recv() {
-                Ok(msg) => msg,
-                Err(_) => {
-                    println!("Collectors disconnected. Stopping exporter.");
-                    return;
-                }
+            let message = match self.receive() {
+                Some(message) => message,
+                None => break,
             };
 
-            match self.exporter.as_ref().unwrap().export(&msg) {
-                Ok(_) => {
-                    println!("Exported message from {} collector", msg.source);
-                }
-                Err(err) => {
-                    println!("Export error: {}", err);
-                    thread::sleep(Duration::from_secs(1));
-                }
+            self.export(message);
+            self.backoff_sleep();
+        }
+
+        println!("Collectors disconnected. Stopping exporter.");
+    }
+
+    fn receive(&mut self) -> Option<Message> {
```

Comment:
Return `Result` here instead of `Option`

Reply:
I will make this change in PR #38